contracts/modules/common/Utils.sol

```diff
@@ -48,6 +57,43 @@ library Utils {
         return recoveredAddress;
     }
 
+    /**
+    * @notice Helper method to recover the spender from a contract call. 
+    * The method returns the contract unless the call is to a standard method of a ERC20/ERC721/ERC1155 token
+    * in which case the spender is recovered from the data.
+    * @param _wallet The calling wallet.
+    * @param _to The target contract.
+    * @param _data The data payload.
+    */
+    function recoverSpender(address _wallet, address _to, bytes memory _data) internal pure returns (address) {
+        if(_data.length >= 4) {
+            bytes4 methodId;
+            // solhint-disable-next-line no-inline-assembly
+            assembly {
+                methodId := mload(add(_data, 0x20))
+            }
+            if(
+                methodId == ERC20_TRANSFER ||
+                methodId == ERC20_APPROVE ||
+                methodId == ERC721_TRANSFER_FROM ||
+                methodId == ERC721_SAFE_TRANSFER_FROM ||
+                methodId == ERC721_SAFE_TRANSFER_FROM_BYTES ||
+                methodId == ERC721_SET_APPROVAL_FOR_ALL ||
+                methodId == ERC1155_SAFE_TRANSFER_FROM
+            ) {
+                address first;
+                address second;
+                // solhint-disable-next-line no-inline-assembly
+                assembly {
+                    first := mload(add(_data, 0x24))
+                    second := mload(add(_data, 0x44))
```

Comment:
We may need to require that `_data.length >= 68`

Reply:
Since the method signature matches one of the standard methods and they all have at least 2 parameters I believe the check is not required.

Reply:
The match is only based on the first 4 bytes of the `_data` and says nothing about the following bytes. You can convince yourself of that by observing that the following test method returns `true` when you call it with just `0xa9059cbb` (no other param).

```
contract T {

    function check(bytes memory _data) public pure returns (bool) {
        bytes4 methodId;
        // solhint-disable-next-line no-inline-assembly
        assembly {
            methodId := mload(add(_data, 0x20))
        }
            
        return bytes4(keccak256("transfer(address,uint256)")) == methodId;
    }
}
```

Reply:
Ha yes, I wasn't considering someone explicitly passing the correct signatures without arguments.

Actually we can then nicely simplify the method:

```
function recoverSpender(address _wallet, address _to, bytes memory _data) internal pure returns (address) {
    if(_data.length >= 68) {
              bytes4 methodId;
              address first;
               address second;
              // solhint-disable-next-line no-inline-assembly
              assembly {
                  methodId := mload(add(_data, 0x20))
                  first := mload(add(_data, 0x24))
                  second := mload(add(_data, 0x44))
              }
              if(
                  methodId == ERC20_TRANSFER ||
                  methodId == ERC20_APPROVE ||
                  methodId == ERC721_TRANSFER_FROM ||
                  methodId == ERC721_SAFE_TRANSFER_FROM ||
                  methodId == ERC721_SAFE_TRANSFER_FROM_BYTES ||
                  methodId == ERC721_SET_APPROVAL_FOR_ALL ||
                  methodId == ERC1155_SAFE_TRANSFER_FROM)
              {
                  return first == _wallet ? second : first;
              }
      }
      return _to;
}
```

Reply:
Fine. Just 66 gas more expensive than doing
```
function recoverSpender(address _wallet, address _to, bytes memory _data) internal pure returns (address) {
    if(_data.length >= 68) {
              bytes4 methodId;
              // solhint-disable-next-line no-inline-assembly
              assembly {
                  methodId := mload(add(_data, 0x20))
              }
              if(
                  methodId == ERC20_TRANSFER ||
                  methodId == ERC20_APPROVE ||
                  methodId == ERC721_TRANSFER_FROM ||
                  methodId == ERC721_SAFE_TRANSFER_FROM ||
                  methodId == ERC721_SAFE_TRANSFER_FROM_BYTES ||
                  methodId == ERC721_SET_APPROVAL_FOR_ALL ||
                  methodId == ERC1155_SAFE_TRANSFER_FROM)
              {
                     address first;
                     address second;
                    // solhint-disable-next-line no-inline-assembly
                    assembly {
                        first := mload(add(_data, 0x24))
                        second := mload(add(_data, 0x44))
                    }
                  return first == _wallet ? second : first;
              }
      }
      return _to;
}
```

Reply:
Damn, no need to waste 66 gas for 3 lines of code.